### PR TITLE
Feature/win lose conditions

### PR DESCRIPTION
**Pull request by lucas1131**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Refactor GridPawn and derivative classes to have a shared CommonGridPawn (blueprint) for shared components and event logic, like HealthBar and `OnDamageTaken` (inheritance now follows GridPawn.cpp -> CommonGridPawn (blueprint) -> specific pawn like PlayerPawn or EnemyAPawn etc. (blueprint);
Implement grayscale healthbar so each pawn can just paint with their own colors independently;
Implement win/lose conditions when all enemies or the player dies.

## Fixes # (issue)
NA

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Source/AmebamanBattleNetjob/BattleManager.h [hunk 1]
```diff
@@ -37,13 +46,13 @@ class AMEBAMANBATTLENETJOB_API ABattleManager : public AActor {
...
+	UPROPERTY(EditDefaultsOnly, Category="Player")
 	FIntVector PlayerGridInitialLocation;
-	UPROPERTY(EditAnywhere, Category="Player")
+	UPROPERTY(EditDefaultsOnly, Category="Player")
 	FIntVector PlayerGridDimensions;
-	UPROPERTY(EditAnywhere, Category="Player")
+	UPROPERTY(EditDefaultsOnly, Category="Player")
 	TSubclassOf<AGrid> PlayerGridBlueprint;
-	UPROPERTY(EditAnywhere, Category="Player")
+	UPROPERTY(EditDefaultsOnly, Category="Player")
```
Comment (g-claudino) [inline Source/AmebamanBattleNetjob/BattleManager.h:55]: Revert to EditAnywhere

File: Source/AmebamanBattleNetjob/BattleManager.cpp [hunk 2]
```diff
@@ -72,21 +72,32 @@ void ABattleManager::PlayerAttackCallback(FIntVector target_offset, int damage){
...
+	if(EnemyGrid->IsPawnInGrid(pawn)){
+		if(EnemyGrid->RemovePawnFromGrid(pawn) <= 0){
+			OnBattleWon();
+		}
+	} else {
+		// do we really need to do this? if player died we'll stop current game flow/logic and go to
+		// a game over flow, so we dont really need to do anything in the grid or destroy the actor?
+		if(PlayerGrid->RemovePawnFromGrid(pawn) <= 0){
+			OnBattleLost();
+		}
```
Comment (g-claudino) [inline Source/AmebamanBattleNetjob/BattleManager.cpp:101]:
```suggestion

```

File: Source/AmebamanBattleNetjob/BattleManager.h [hunk 3]
```diff
@@ -52,21 +61,16 @@ class AMEBAMANBATTLENETJOB_API ABattleManager : public AActor {
...
+	UPROPERTY(EditDefaultsOnly, Category="Enemies")
 	TArray<FIntVector> EnemiesGridInitialLocation;
-	UPROPERTY(EditAnywhere, Category="Enemies")
+	UPROPERTY(EditDefaultsOnly, Category="Enemies")
 	FIntVector EnemyGridDimensions;
-	UPROPERTY(EditAnywhere, Category="Enemies")
+	UPROPERTY(EditDefaultsOnly, Category="Enemies")
 	TSubclassOf<AGrid> EnemyGridBlueprint;
-	UPROPERTY(EditAnywhere, Category="Enemies")
+	UPROPERTY(EditDefaultsOnly, Category="Enemies")
```
Comment (g-claudino) [inline Source/AmebamanBattleNetjob/BattleManager.h:70]: Revert to EditAnywhere

File: Source/AmebamanBattleNetjob/Grid.cpp [hunk 4]
```diff
@@ -160,6 +160,10 @@ inline FIntVector AGrid::GridArrayIndexToFIntVector(int idx) {
     return FIntVector(x, y, z);
 }
 
+inline bool AGrid::IsPawnInGrid(AGridPawn *pawn){
+	return GridPawnMap.Find(pawn->GetName()) != nullptr;
```
Comment (g-claudino) [inline Source/AmebamanBattleNetjob/Grid.cpp:164]: Use is valid

**Review by g-claudino — CHANGES_REQUESTED** [review]
Great bro, just fix a few things as they're in the specific comments

File: Source/AmebamanBattleNetjob/Grid.cpp (hunk 4 above)
Comment (lucas1131) [reply Source/AmebamanBattleNetjob/Grid.cpp:164]: In this case we are returning a `struct` and not a `UObject` so we cant use `IsValid`. Also, find returns explicitly `nullptr` if key was not found in the map.

**Review by g-claudino — APPROVED** [review]
Bom demais

**Pull request merged**